File: discrete-math/labs/lab-6/A.cpp
/*
 * Nariman Safiulin (woofilee)
 * File: A.cpp
 * Created on: Dec 25, 2016
 */

#include <bits/stdc++.h>

#define TASK "rainbow"

using namespace std;

ifstream in(TASK ".in");
ofstream out(TASK ".ans");

#define pb push_back
#define N 100
#define M 5000

struct Edge {
    int u;
    int v;
    int color;
};

int n;
int m;

vector<Edge> edges;
vector<int> graph[M];

set<int> right_set;
set<int> result_set;

bool color[N];
bool used[M];
bool x1[M];
bool x2[M];

int parent[M];

int msu_rank[N];
int msu[N];

void reset() {
    for (int i = 0; i < n; msu[i] = i, i++);
}

int get(int v) {
    return (v == msu[v]) ? msu[v] : msu[v] = get(msu[v]);
}

void unite(int u, int v) {
    int a = get(u);
    int b = get(v);

    if (a == b)
        return;

    if (msu_rank[a] < msu_rank[b])
        swap(a, b);

    if (msu_rank[a] == msu_rank[b])
        msu_rank[a]++;

    msu[b] = a;
}

int main() {
    in >> n >> m;

    for (int i = 0; i < m; right_set.insert(i++));

    for (int i = 0; i < m; i++) {
        int a, b, c;
        in >> a >> b >> c;
        edges.pb({ --a, --b, --c });
    }

    while(true) {
        for (int i = 0; i < m; graph[i++].clear());

        for (auto a : result_set) {
            reset();

            for (auto b : result_set)
                if (b == a)
                    continue;
                else
                    unite(edges[b].u, edges[b].v);

            for (auto c : right_set) {
                if (get(edges[c].u) != get(edges[c].v))
                    graph[a].pb(c);
                if (edges[c].color == edges[a].color || !color[edges[c].color])
                    graph[c].pb(a);
            }
        }

        reset();
        for (int i = 0; i < m; x1[i] = false, x2[i++] = false);

        for (auto l : result_set)
             unite(edges[l].u, edges[l].v);

        for (auto l : right_set) {
            if (get(edges[l].u) != get(edges[l].v))
                x1[l] = true;
            if (!color[edges[l].color])
                x2[l] = true;
        }

        queue<int> queue;

        for (int i = 0; i < m; i++) {
            used[i] = false;

            if (x1[i]) {
                parent[i] = -1;
                used[i] = true;
                queue.push(i);
            }
        }

        while (!queue.empty()) {
            int l = queue.front(); queue.pop();

            if (x2[l]) {
                while (l != -1) {
                    if (right_set.count(l) > 0) {
                        result_set.insert(l);
                        right_set.erase(l);
                        color[edges[l].color] = true;
                    } else {
                        result_set.erase(l);
                        right_set.insert(l);
                        color[edges[l].color] = false;
                    }

                    l = parent[l];
                }

                goto endloop;
            }

            for (auto k : graph[l])
                if (!used[k]) {
                    parent[k] = l;
                    used[k] = true;
                    queue.push(k);
                }
        }

        break;
        endloop:;
    }

    out << result_set.size() << "\n";
    for (auto l : result_set)
        out << l + 1 << " ";

    return 0;
}

File: discrete-math/labs/lab-6/A.kt
/*
 * Nariman Safiulin (woofilee)
 * File: A.kt
 * Created on: Dec 24, 2016
 */

import java.io.*
import java.util.*
import java.nio.file.*

private val PROBLEM_NAME = "rainbow"

// Aarrgghh... This solution is highly optimized, but runs to Time Limit on test 7 :(

private class Scanner(f: File) {
    val r = f.bufferedReader()
    var t = StringTokenizer("")

    fun hasNext(): Boolean {
        while (!t.hasMoreTokens()) t = StringTokenizer(r.readLine() ?: return false)
        return true
    }

    fun next(): String = if (hasNext()) t.nextToken() else ""
    fun int(): Int = next().toInt()
    fun long(): Long = next().toLong()
    fun double(): Double = next().toDouble()
    fun close() = r.close()
}

data class Edge(val u: Int, val v: Int, val color: Int)

private fun solve(`in`: Scanner, out: BufferedWriter) {
    val n = `in`.int()
    val m = `in`.int()

    val edges = Array(m) { Edge(`in`.int() - 1, `in`.int() - 1, `in`.int() - 1) }
    val graph = Array(m) { Array(m) { 0 } }
    val graphSize = Array(m) { 0 }

    val right = HashSet<Int>()
    val result = HashSet<Int>()

    for (i in 0..m - 1) right.add(i)

    val color = Array(100) { false }
    val used = Array(m) { false }
    val x1 = Array(m) { false }
    val x2 = Array(m) { false }

    val parent = Array(m) { 0 }

    val msu = Array(n) { it }
    val rank = Array(n) { 0 }

    val queue = Array(m) { 0 }

    fun get(v: Int): Int {
        if (v != msu[v]) msu[v] = get(msu[v])
        return msu[v]
    }

    fun unite(u: Int, v: Int) {
        var a = get(u)
        var b = get(v)

        if (a == b) return

        if (rank[a] < rank[b]) {
            val t = a
            a = b
            b = t
        }

        if (rank[a] == rank[b]) rank[a]++

        msu[b] = a
    }

    out@ while (true) {
        for (i in 0..m - 1) graphSize[i] = 0

        val re: Array<Int> = result.toTypedArray()
        val ri: Array<Int> = right.toTypedArray()

        for (ai in 0..re.size - 1) {
            val a = re[ai]

            for (i in 0..n - 1) msu[i] = i
            for (bi in 0..re.size - 1)  if (bi == ai) continue else unite(edges[re[bi]].u, edges[a].v)

            for (ci in 0..ri.size - 1) {
                val c = ri[ci]
                if (get(edges[c].u) != get(edges[c].v)) graph[a][graphSize[a]++] = c
                if (edges[c].color == edges[a].color || !color[edges[c].color]) graph[c][graphSize[c]++] = a
            }
        }

        for (i in 0..n - 1) msu[i] = i
        for (i in 0..m - 1) {
            x1[i] = false
            x2[i] = false
        }

        for (li in 0..re.size - 1) unite(edges[re[li]].u, edges[re[li]].v)
        for (li in 0..ri.size - 1) {
            val l = ri[li]
            if (get(edges[l].u) != get(edges[l].v)) x1[l] = true
            if (!color[edges[l].color]) x2[l] = true
        }

        var queueStart = 0
        var queueEnd = 0

        for (i in 0..m - 1) {
            used[i] = false

            if (x1[i]) {
                parent[i] = -1
                used[i] = true
                queue[queueEnd++] = i
            }
        }

        while (queueStart != queueEnd) {
            var l = queue[queueStart++]

            if (x2[l]) {
                while (l != -1) {
                    if (right.contains(l)) {
                        result.add(l)
                        right.remove(l)
                        color[edges[l].color] = true
                    } else {
                        result.remove(l)
                        right.add(l)
                        color[edges[l].color] = false
                    }

                    l = parent[l]
                }

                continue@out
            }

            for (i in 0..graphSize[l] - 1) {
                val k = graph[l][i]

                if (!used[k]) {
                    parent[k] = l
                    used[k] = true
                    queue[queueEnd++] = k
                }
            }
        }

        break
    }

    out.write("${result.size}\n")
    for (it in result) out.write("${it + 1} ")
}

private fun naive(`in`: Scanner, out: BufferedWriter) {
    Runtime.getRuntime().exec("cmd /c A.exe").waitFor()
}

private fun testgen(`in`: Scanner, out: BufferedWriter) {
    val random = Random()
    val n = random.nextInt(1) + 100
    val m = random.nextInt(1) + 5000

    out.write("$n $m\n")

    for (i in 1..m) {
        val a = random.nextInt(n) + 1
        var b = random.nextInt(n) + 1
        val c = random.nextInt(100) + 1

        while (a == b) b = random.nextInt(n) + 1

        out.write("$a $b $c\n")
    }

    print("N=$n, M=$m... ")
}

private fun run(name: String, ie: String, oe: String, f: (Scanner, BufferedWriter) -> Unit) {
    print("$name running... ")
    val start = System.nanoTime()

    val `in` = Scanner(File(PROBLEM_NAME + ".$ie"))
    val out = File(PROBLEM_NAME + ".$oe").bufferedWriter()

    f(`in`, out)

    `in`.close()
    out.close()

    val end = System.nanoTime()
    println("OK, ${(end - start) / 1e6} ms.")
}

fun main(args: Array<String>) {
    val judge = false
    val naive = false
    val tests = 150

    val ie = "in"
    val ae = "ans"
    val oe = "out"

    if (!judge) {
        run("Solution", ie, oe, ::solve)
    } else {
        repeat(tests) {
            println("Checking test #${it + 1}...")

            print("  - "); run("Test generation", oe, ie, ::testgen)
            if (naive) { print("  - "); run("Naive solution", ie, ae, ::naive) }
            print("  - "); run("Main solution", ie, oe, ::solve)

            if (naive) {
                print("  - Checking... ")

                if (Files.readAllLines(Paths.get(PROBLEM_NAME + ".out"))[0] ==
                        Files.readAllLines(Paths.get(PROBLEM_NAME + ".ans"))[0]) {
                    println("OK")
                } else {
                    println("WA")
                    return
                }
            }

            println()
        }
    }
}
